Missing information when Error creating Mimir session (Caused by: java.net.ConnectException: Connection refused)

Stacktrace:

```
[ERROR] Error creating Mimir session: Connection refused -> [Help 1]
org.apache.maven.MavenExecutionException: Error creating Mimir session
    at eu.maveniverse.maven.mimir.extension3.MimirLifecycleParticipant.afterProjectsRead (MimirLifecycleParticipant.java:73)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:223)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: java.net.ConnectException: Connection refused
    at sun.nio.ch.UnixDomainSockets.connect0 (Native Method)
    at sun.nio.ch.UnixDomainSockets.connect (UnixDomainSockets.java:154)
    at sun.nio.ch.UnixDomainSockets.connect (UnixDomainSockets.java:150)
    at sun.nio.ch.SocketChannelImpl.connect (SocketChannelImpl.java:878)
    at eu.maveniverse.maven.mimir.daemon.protocol.Handle$2.getHandle (Handle.java:97)
    at eu.maveniverse.maven.mimir.node.daemon.DaemonNode.<init> (DaemonNode.java:56)
    at eu.maveniverse.maven.mimir.node.daemon.DaemonNodeFactory.createNode (DaemonNodeFactory.java:55)
    at eu.maveniverse.maven.mimir.node.daemon.DaemonNodeFactory.createNode (DaemonNodeFactory.java:28)
    at eu.maveniverse.maven.mimir.shared.impl.SessionFactoryImpl.createSession (SessionFactoryImpl.java:62)
    at eu.maveniverse.maven.mimir.extension3.MimirLifecycleParticipant.afterProjectsRead (MimirLifecycleParticipant.java:68)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:223)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)

```

The fix is as simple as:
```
rm -rf ~/.mimir/mimir-socket
```

What do you think about giving information about the local path and tell user to check it?

--- a/daemon-protocol/src/main/java/eu/maveniverse/maven/mimir/daemon/protocol/Handle.java
+++ b/daemon-protocol/src/main/java/eu/maveniverse/maven/mimir/daemon/protocol/Handle.java
@@ -20,6 +20,7 @@ import java.nio.channels.ByteChannel;
 import java.nio.channels.Channels;
 import java.nio.channels.ServerSocketChannel;
 import java.nio.channels.SocketChannel;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.LinkedHashMap;
 import java.util.Map;
@@ -65,7 +66,11 @@ public class Handle implements Closeable {
             @Override
             public void close() throws IOException {
                 if (closed.compareAndSet(false, true)) {
-                    serverSocketChannel.close();
+                    try {
+                        serverSocketChannel.close();
+                    } finally {
+                        Files.deleteIfExists(domainSocketPath);
+                    }
                 }
             }
         };

--- a/daemon/src/main/java/eu/maveniverse/maven/mimir/daemon/Daemon.java
+++ b/daemon/src/main/java/eu/maveniverse/maven/mimir/daemon/Daemon.java
@@ -75,6 +75,7 @@ public class Daemon extends CloseableConfigSupport<DaemonConfig> implements Clos
             Runtime.getRuntime().addShutdownHook(new Thread(daemon::shutdown));
         } catch (Exception e) {
             e.printStackTrace(System.out);
+            System.exit(1);
         }
     }
 

--- a/daemon/src/main/java/eu/maveniverse/maven/mimir/daemon/Daemon.java
+++ b/daemon/src/main/java/eu/maveniverse/maven/mimir/daemon/Daemon.java
@@ -19,7 +19,8 @@ import eu.maveniverse.maven.mimir.shared.node.RemoteNode;
 import eu.maveniverse.maven.mimir.shared.node.RemoteNodeFactory;
 import eu.maveniverse.maven.mimir.shared.node.SystemNode;
 import eu.maveniverse.maven.mimir.shared.node.SystemNodeFactory;
-import eu.maveniverse.maven.shared.core.component.ComponentSupport;
+import eu.maveniverse.maven.shared.core.component.CloseableConfigSupport;
+import eu.maveniverse.maven.shared.core.fs.DirectoryLocker;
 import java.io.Closeable;
 import java.io.IOException;
 import java.nio.channels.AsynchronousCloseException;
@@ -47,7 +48,7 @@ import org.eclipse.sisu.wire.WireModule;
 
 @Named
 @Singleton
-public class Daemon extends ComponentSupport implements Closeable {
+public class Daemon extends CloseableConfigSupport<DaemonConfig> implements Closeable {
     static {
         // make Slf4j-simple go for stdout and not default stderr (unless re-configured by user)
         if (System.getProperty("org.slf4j.simpleLogger.logFile") == null) {
@@ -64,7 +65,6 @@ public class Daemon extends ComponentSupport implements Closeable {
                             new AbstractModule() {
                                 @Override
                                 protected void configure() {
-                                    bind(Config.class).toInstance(config);
                                     bind(DaemonConfig.class).toInstance(daemonConfig);
                                 }
                             },
@@ -72,7 +72,7 @@ public class Daemon extends ComponentSupport implements Closeable {
                                     new URLClassSpace(Daemon.class.getClassLoader()), BeanScanning.INDEX, true)))
                     .getInstance(Daemon.class);
 
-            Runtime.getRuntime().addShutdownHook(new Thread(daemon::close));
+            Runtime.getRuntime().addShutdownHook(new Thread(daemon::shutdown));
         } catch (Exception e) {
             e.printStackTrace(System.out);
         }
@@ -83,15 +83,14 @@ public class Daemon extends ComponentSupport implements Closeable {
         private final SystemNode<?> systemNode;
 
         @Inject
-        public SystemNodeProvider(
-                Config config, DaemonConfig daemonConfig, Map<String, SystemNodeFactory> systemNodeFactories)
+        public SystemNodeProvider(DaemonConfig daemonConfig, Map<String, SystemNodeFactory> systemNodeFactories)
                 throws IOException {
             requireNonNull(systemNodeFactories, "systemNodeFactories");
             SystemNodeFactory systemNodeFactory = systemNodeFactories.get(daemonConfig.systemNode());
             if (systemNodeFactory == null) {
                 throw new IllegalArgumentException("Unknown system node: " + daemonConfig.systemNode());
             }
-            this.systemNode = systemNodeFactory.createNode(config);
+            this.systemNode = systemNodeFactory.createNode(daemonConfig.config());
         }
 
         @Override
@@ -100,30 +99,25 @@ public class Daemon extends ComponentSupport implements Closeable {
         }
     }
 
-    private final Config config;
-    private final DaemonConfig daemonConfig;
-    private final Handle.ServerHandle serverHandle;
     private final ExecutorService executor;
     private final SystemNode<?> systemNode;
     private final List<RemoteNode<?>> remoteNodes;
+    private final Handle.ServerHandle serverHandle;
 
     @Inject
     public Daemon(
-            Config config,
             DaemonConfig daemonConfig,
             SystemNode<?> systemNode,
             Map<String, RemoteNodeFactory> remoteNodeFactories,
             Map<String, ChecksumAlgorithmFactory> checksumAlgorithmFactories)
             throws IOException {
-        this.config = requireNonNull(config, "config");
-        this.daemonConfig = requireNonNull(daemonConfig, "daemonConfig");
-        requireNonNull(systemNode, "systemNode");
+        super(daemonConfig);
+        this.systemNode = requireNonNull(systemNode, "systemNode");
         requireNonNull(remoteNodeFactories, "remoteNodeFactories");
 
-        this.systemNode = systemNode;
         ArrayList<RemoteNode<?>> nds = new ArrayList<>();
         for (RemoteNodeFactory remoteNodeFactory : remoteNodeFactories.values()) {
-            Optional<? extends RemoteNode<?>> node = remoteNodeFactory.createNode(config);
+            Optional<? extends RemoteNode<?>> node = remoteNodeFactory.createNode(config.config());
             node.ifPresent(nds::add);
         }
         nds.sort(Comparator.comparing(RemoteNode::distance));
@@ -131,20 +125,17 @@ public class Daemon extends ComponentSupport implements Closeable {
         // Java 21: this.executor = Executors.newVirtualThreadPerTaskExecutor();
         this.executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() - 1);
 
+        // lock exclusively the basedir; if other daemon tries to run here will fail
+        Files.createDirectories(config.daemonBasedir());
+        DirectoryLocker.INSTANCE.lockDirectory(config.daemonBasedir(), true);
+
         Path socketPath = daemonConfig.socketPath();
-        // make sure socket is deleted once daemon exits
-        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
-            try {
-                Files.deleteIfExists(socketPath);
-            } catch (IOException e) {
-                logger.warn("Failed to delete socket path: {}", socketPath, e);
-            }
-        }));
+        Files.deleteIfExists(socketPath);
         this.serverHandle = Handle.serverDomainSocket(socketPath);
 
-        logger.info("Mimir Daemon {} started", config.mimirVersion().orElse("UNKNOWN"));
+        logger.info("Mimir Daemon {} started", config.config().mimirVersion().orElse("UNKNOWN"));
         logger.info("  PID: {}", ProcessHandle.current().pid());
-        logger.info("  Properties: {}", config.basedir().resolve(config.propertiesPath()));
+        logger.info("  Properties: {}", config.config().propertiesPath());
         logger.info("  Supported checksums: {}", checksumAlgorithmFactories.keySet());
         logger.info("  Socket: {}", socketPath);
         logger.info("  System Node: {}", systemNode);
@@ -156,7 +147,7 @@ public class Daemon extends ComponentSupport implements Closeable {
 
         HashMap<String, String> daemonData = new HashMap<>();
         daemonData.put(Session.DAEMON_PID, Long.toString(ProcessHandle.current().pid()));
-        daemonData.put(Session.DAEMON_VERSION, config.mimirVersion().orElse("UNKNOWN"));
+        daemonData.put(Session.DAEMON_VERSION, config.config().mimirVersion().orElse("UNKNOWN"));
 
         Predicate<Request> clientPredicate =
                 req -> Objects.equals(req.requireData(Session.NODE_VERSION), daemonData.get(Session.DAEMON_VERSION));
@@ -166,7 +157,7 @@ public class Daemon extends ComponentSupport implements Closeable {
                 while (serverHandle.isOpen()) {
                     Handle handle = serverHandle.accept();
                     executor.submit(new DaemonServer(
-                            handle, daemonData, systemNode, remoteNodes, clientPredicate, this::close));
+                            handle, daemonData, systemNode, remoteNodes, clientPredicate, this::shutdown));
                 }
             } catch (AsynchronousCloseException ignored) {
                 // we are done
@@ -176,8 +167,16 @@ public class Daemon extends ComponentSupport implements Closeable {
         });
     }
 
+    public void shutdown() {
+        try {
+            close();
+        } catch (IOException e) {
+            logger.warn("Failed to close daemon", e);
+        }
+    }
+
     @Override
-    public void close() {
+    protected void doClose() throws IOException {
         try {
             try {
                 serverHandle.close();
@@ -202,6 +201,7 @@ public class Daemon extends ComponentSupport implements Closeable {
                 logger.warn("Error closing local node", e);
             }
         } finally {
+            DirectoryLocker.INSTANCE.unlockDirectory(config.daemonBasedir());
             logger.info("Daemon stopped");
         }
     }

--- a/daemon/src/main/java/eu/maveniverse/maven/mimir/daemon/DaemonConfig.java
+++ b/daemon/src/main/java/eu/maveniverse/maven/mimir/daemon/DaemonConfig.java
@@ -17,6 +17,7 @@ public class DaemonConfig {
     public static DaemonConfig with(Config config) {
         requireNonNull(config, "config");
 
+        Path daemonBasedir = config.basedir().resolve("daemon");
         Path socketPath = config.basedir().resolve(Handle.DEFAULT_SOCKET_PATH);
         String systemNode = "file";
 
@@ -27,15 +28,27 @@ public class DaemonConfig {
         if (config.effectiveProperties().containsKey("mimir.daemon.systemNode")) {
             systemNode = config.effectiveProperties().get("mimir.daemon.systemNode");
         }
-        return new DaemonConfig(socketPath, systemNode);
+        return new DaemonConfig(config, daemonBasedir, socketPath, systemNode);
     }
 
+    private final Config config;
+    private final Path daemonBasedir;
     private final Path socketPath;
     private final String systemNode;
 
-    private DaemonConfig(Path socketPath, String systemNode) {
+    private DaemonConfig(Config config, Path daemonBasedir, Path socketPath, String systemNode) {
+        this.config = requireNonNull(config);
+        this.daemonBasedir = requireNonNull(daemonBasedir);
         this.socketPath = requireNonNull(socketPath);
         this.systemNode = requireNonNull(systemNode);
+    }
+
+    public Config config() {
+        return config;
+    }
+
+    public Path daemonBasedir() {
+        return daemonBasedir;
     }
 
     public Path socketPath() {

--- a/extension/src/main/java/eu/maveniverse/maven/mimir/extension3/MimirLifecycleParticipant.java
+++ b/extension/src/main/java/eu/maveniverse/maven/mimir/extension3/MimirLifecycleParticipant.java
@@ -12,7 +12,6 @@ import eu.maveniverse.maven.mimir.shared.Config;
 import eu.maveniverse.maven.mimir.shared.SessionFactory;
 import eu.maveniverse.maven.shared.core.fs.FileUtils;
 import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.List;
 import java.util.Objects;
 import javax.inject.Inject;
@@ -92,8 +91,7 @@ public class MimirLifecycleParticipant extends AbstractMavenLifecycleParticipant
             logger.debug("Not resolving Mimir daemon; autostart not enabled or version not detected");
             return;
         }
-        Path daemonJarPath = config.basedir().resolve(daemonConfig.daemonJarName());
-        if (!Files.exists(daemonJarPath)) {
+        if (!Files.exists(daemonConfig.daemonJar())) {
             try {
                 logger.info(
                         "Resolving Mimir daemon version {}",
@@ -101,8 +99,8 @@ public class MimirLifecycleParticipant extends AbstractMavenLifecycleParticipant
                 ArtifactRequest artifactRequest =
                         new ArtifactRequest(new DefaultArtifact(daemonConfig.daemonGav()), remoteRepositories, "mimir");
                 ArtifactResult artifactResult = repositorySystem.resolveArtifact(session, artifactRequest);
-                Files.createDirectories(daemonJarPath.getParent());
-                FileUtils.copyOrLink(artifactResult.getArtifact().getFile().toPath(), daemonJarPath);
+                Files.createDirectories(daemonConfig.daemonJar().getParent());
+                FileUtils.copyOrLink(artifactResult.getArtifact().getFile().toPath(), daemonConfig.daemonJar());
             } catch (Exception e) {
                 logger.warn("Failed to resolve daemon: {}", e.getMessage());
             }

--- a/node/daemon/src/main/java/eu/maveniverse/maven/mimir/node/daemon/DaemonConfig.java
+++ b/node/daemon/src/main/java/eu/maveniverse/maven/mimir/node/daemon/DaemonConfig.java
@@ -21,6 +21,7 @@ public class DaemonConfig {
         final boolean mimirVersionPresent = config.mimirVersion().isPresent();
         final String mimirVersion = config.mimirVersion().orElse("UNKNOWN");
 
+        Path daemonBasedir = config.basedir().resolve("daemon");
         Path socketPath = config.basedir().resolve(Handle.DEFAULT_SOCKET_PATH);
         Path daemonJavaHome = Path.of(config.effectiveProperties()
                 .getOrDefault(
@@ -29,8 +30,8 @@ public class DaemonConfig {
         boolean autostart = mimirVersionPresent;
         Duration autostartDuration = Duration.ofMinutes(1);
         boolean autostop = false;
-        String daemonJarName = "daemon-" + mimirVersion + ".jar";
-        String daemonLogName = "daemon-" + mimirVersion + ".log";
+        Path daemonJar = config.basedir().resolve("daemon-" + mimirVersion + ".jar");
+        Path daemonLog = config.basedir().resolve("daemon-" + mimirVersion + ".log");
         String daemonGav = "eu.maveniverse.maven.mimir:daemon:jar:daemon:" + mimirVersion;
         boolean passOnBasedir = false;
         boolean debug = false;
@@ -51,11 +52,11 @@ public class DaemonConfig {
         if (config.effectiveProperties().containsKey("mimir.daemon.autostop")) {
             autostop = Boolean.parseBoolean(config.effectiveProperties().get("mimir.daemon.autostop"));
         }
-        if (config.effectiveProperties().containsKey("mimir.daemon.daemonJarName")) {
-            daemonJarName = config.effectiveProperties().get("mimir.daemon.daemonJarName");
+        if (config.effectiveProperties().containsKey("mimir.daemon.daemonJar")) {
+            daemonJar = config.basedir().resolve(config.effectiveProperties().get("mimir.daemon.daemonJar"));
         }
-        if (config.effectiveProperties().containsKey("mimir.daemon.daemonLogName")) {
-            daemonLogName = config.effectiveProperties().get("mimir.daemon.daemonLogName");
+        if (config.effectiveProperties().containsKey("mimir.daemon.daemonLog")) {
+            daemonLog = config.basedir().resolve(config.effectiveProperties().get("mimir.daemon.daemonLog"));
         }
         if (config.effectiveProperties().containsKey("mimir.daemon.daemonGav")) {
             daemonGav = config.effectiveProperties().get("mimir.daemon.daemonGav");
@@ -67,14 +68,16 @@ public class DaemonConfig {
             debug = Boolean.parseBoolean(config.effectiveProperties().get("mimir.daemon.debug"));
         }
         return new DaemonConfig(
+                config,
+                daemonBasedir,
                 socketPath,
                 daemonJavaHome,
                 autoupdate,
                 autostart,
                 autostartDuration,
                 autostop,
-                daemonJarName,
-                daemonLogName,
+                daemonJar,
+                daemonLog,
                 daemonGav,
                 passOnBasedir,
                 debug);
@@ -82,41 +85,55 @@ public class DaemonConfig {
 
     public static final String NAME = "daemon";
 
+    private final Config config;
+    private final Path daemonBasedir;
     private final Path socketPath;
     private final Path daemonJavaHome;
     private final boolean autoupdate;
     private final boolean autostart;
     private final Duration autostartDuration;
     private final boolean autostop;
-    private final String daemonJarName;
-    private final String daemonLogName;
+    private final Path daemonJar;
+    private final Path daemonLog;
     private final String daemonGav;
     private final boolean passOnBasedir;
     private final boolean debug;
 
     private DaemonConfig(
+            Config config,
+            Path daemonBasedir,
             Path socketPath,
             Path daemonJavaHome,
             boolean autoupdate,
             boolean autostart,
             Duration autostartDuration,
             boolean autostop,
-            String daemonJarName,
-            String daemonLogName,
+            Path daemonJar,
+            Path daemonLog,
             String daemonGav,
             boolean passOnBasedir,
             boolean debug) {
+        this.config = requireNonNull(config);
+        this.daemonBasedir = requireNonNull(daemonBasedir);
         this.socketPath = requireNonNull(socketPath);
         this.daemonJavaHome = requireNonNull(daemonJavaHome);
         this.autoupdate = autoupdate;
         this.autostart = autostart;
         this.autostartDuration = requireNonNull(autostartDuration);
         this.autostop = autostop;
-        this.daemonJarName = requireNonNull(daemonJarName);
-        this.daemonLogName = requireNonNull(daemonLogName);
+        this.daemonJar = requireNonNull(daemonJar);
+        this.daemonLog = requireNonNull(daemonLog);
         this.daemonGav = requireNonNull(daemonGav);
         this.passOnBasedir = passOnBasedir;
         this.debug = debug;
+    }
+
+    public Config config() {
+        return config;
+    }
+
+    public Path daemonBasedir() {
+        return daemonBasedir;
     }
 
     public Path socketPath() {
@@ -143,12 +160,12 @@ public class DaemonConfig {
         return autostop;
     }
 
-    public String daemonJarName() {
-        return daemonJarName;
+    public Path daemonJar() {
+        return daemonJar;
     }
 
-    public String daemonLogName() {
-        return daemonLogName;
+    public Path daemonLog() {
+        return daemonLog;
     }
 
     public String daemonGav() {

--- a/node/daemon/src/main/java/eu/maveniverse/maven/mimir/node/daemon/DaemonNodeFactory.java
+++ b/node/daemon/src/main/java/eu/maveniverse/maven/mimir/node/daemon/DaemonNodeFactory.java
@@ -13,6 +13,7 @@ import eu.maveniverse.maven.mimir.daemon.protocol.Session;
 import eu.maveniverse.maven.mimir.shared.Config;
 import eu.maveniverse.maven.mimir.shared.node.LocalNodeFactory;
 import eu.maveniverse.maven.shared.core.component.ComponentSupport;
+import eu.maveniverse.maven.shared.core.fs.DirectoryLocker;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -38,15 +39,24 @@ public class DaemonNodeFactory extends ComponentSupport implements LocalNodeFact
     @Override
     public DaemonNode createNode(Config config) throws IOException {
         DaemonConfig cfg = DaemonConfig.with(config);
-        if (!Files.exists(cfg.socketPath())) {
+        if (tryLock(cfg)) {
+            Files.deleteIfExists(cfg.socketPath());
             if (cfg.autostart()) {
                 logger.debug("Mimir daemon is not running, starting it");
-                Process daemon = startDaemon(config.basedir(), config, cfg);
-                if (daemon == null) {
-                    throw new IOException("Mimir daemon could not be started");
-                }
+                Process daemon = startDaemon(cfg);
                 logger.info("Mimir daemon started (pid={})", daemon.pid());
+            } else {
+                throw new IOException("Mimir daemon does not run and autostart is disabled; start daemon manually");
             }
+        } else {
+            if (!Files.exists(cfg.socketPath())) {
+                waitForSocket(cfg);
+            }
+        }
+
+        // at this point socket must exist
+        if (!Files.exists(cfg.socketPath())) {
+            throw new IOException("Mimir daemon socket not found");
         }
         HashMap<String, String> clientData = new HashMap<>();
         clientData.put(Session.NODE_PID, Long.toString(ProcessHandle.current().pid()));
@@ -54,70 +64,122 @@ public class DaemonNodeFactory extends ComponentSupport implements LocalNodeFact
         try {
             return new DaemonNode(clientData, cfg.socketPath(), checksumAlgorithmFactories, cfg.autostop());
         } catch (IOException e) {
-            mayDumpDaemonLog(config.basedir().resolve(cfg.daemonLogName()));
+            mayDumpDaemonLog(config.basedir().resolve(cfg.daemonLog()));
             throw e;
         }
     }
 
-    private Process startDaemon(Path basedir, Config config, DaemonConfig daemonConfig) throws IOException {
-        String daemonJarName = daemonConfig.daemonJarName();
-        String daemonLogName = daemonConfig.daemonLogName();
-        Path daemonJar = basedir.resolve(daemonJarName);
-        Path daemonLog = basedir.resolve(daemonLogName);
-        if (Files.isRegularFile(daemonJar)) {
-            String java = daemonConfig
-                    .daemonJavaHome()
+    /**
+     * Starts damon process. This method must be entered ONLY if caller owns exclusive lock "start procedure".
+     *
+     * @see #tryLock(DaemonConfig)
+     */
+    private Process startDaemon(DaemonConfig cfg) throws IOException {
+        Path basedir = cfg.config().basedir();
+        if (Files.isRegularFile(cfg.daemonJar())) {
+            String java = cfg.daemonJavaHome()
                     .resolve("bin")
                     .resolve(
-                            config.effectiveProperties()
+                            cfg.config()
+                                            .effectiveProperties()
                                             .getOrDefault("os.name", "unknown")
                                             .startsWith("Windows")
                                     ? "java.exe"
                                     : "java")
                     .toString();
-            ProcessBuilder pb = new ProcessBuilder().directory(basedir.toFile()).redirectOutput(daemonLog.toFile());
 
             ArrayList<String> command = new ArrayList<>();
             command.add(java);
-            if (daemonConfig.debug()) {
+            if (cfg.debug()) {
                 command.add("-Dorg.slf4j.simpleLogger.defaultLogLevel=debug");
             }
-            if (daemonConfig.passOnBasedir()) {
+            if (cfg.passOnBasedir()) {
                 command.add("-Dmimir.basedir=" + basedir);
             }
             command.add("-jar");
-            command.add(daemonJarName);
+            command.add(cfg.daemonJar().toString());
 
-            pb.command(command);
+            ProcessBuilder pb = new ProcessBuilder()
+                    .directory(basedir.toFile())
+                    .redirectOutput(cfg.daemonLog().toFile())
+                    .command(command);
+
+            unlock(cfg);
+
             Process p = pb.start();
-            Instant startingUntil = Instant.now().plus(daemonConfig.autostartDuration());
             try {
-                while (p.isAlive() && !Files.exists(daemonConfig.socketPath())) {
-                    if (Instant.now().isAfter(startingUntil)) {
-                        mayDumpDaemonLog(daemonLog);
-                        throw new IOException("Failed to start daemon in time " + daemonConfig.autostartDuration()
-                                + "; check daemon logs in " + daemonConfig.daemonLogName());
-                    }
-                    logger.debug("Waiting for socket to open");
-                    Thread.sleep(500);
-                }
-            } catch (InterruptedException e) {
-                throw new IOException("Interrupted", e);
+                waitForSocket(cfg);
+            } catch (IOException e) {
+                p.destroy();
+                throw e;
             }
             if (p.isAlive()) {
                 return p;
             } else {
-                mayDumpDaemonLog(daemonLog);
-                throw new IOException("Failed to start daemon; check daemon logs in " + daemonConfig.daemonLogName());
+                mayDumpDaemonLog(cfg.daemonLog());
+                throw new IOException("Failed to start daemon; check daemon logs in " + cfg.daemonLog());
             }
+        } else {
+            throw new IOException("Mimir daemon JAR not found");
         }
-        logger.warn("Mimir daemon is not present; cannot start it");
-        return null;
     }
 
+    /**
+     * Dumps the daemon log file for user.
+     */
     private void mayDumpDaemonLog(Path daemonLog) throws IOException {
         if (Files.isRegularFile(daemonLog)) {
             logger.error("Daemon log dump:\n{}", Files.readString(daemonLog));
+        }
+    }
+
+    /**
+     * Locks the {@link DaemonConfig#daemonBasedir()}. If this method returns {@code true} it means there is no
+     * daemon running nor is there any other process trying to start daemon.
+     * This process "owns" the start procedure alone.
+     */
+    private boolean tryLock(DaemonConfig cfg) {
+        try {
+            Files.createDirectories(cfg.daemonBasedir());
+            DirectoryLocker.INSTANCE.lockDirectory(cfg.daemonBasedir(), true);
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Unlocks the {@link DaemonConfig#daemonBasedir()}.
+     */
+    private void unlock(DaemonConfig cfg) throws IOException {
+        DirectoryLocker.INSTANCE.unlockDirectory(cfg.daemonBasedir());
+    }
+
+    /**
+     * The method will wait {@link DaemonConfig#autostartDuration()} time for socket to become available.
+     * Precondition: socket does not exist.
+     * Exit condition: socket exist.
+     * Fail condition: time passes and socket not exist.
+     */
+    private void waitForSocket(DaemonConfig cfg) throws IOException {
+        Instant startingUntil = Instant.now().plus(cfg.autostartDuration());
+        logger.debug("Waiting for socket to become available until {}", startingUntil);
+        try {
+            while (!Files.exists(cfg.socketPath())) {
+                if (Instant.now().isAfter(startingUntil)) {
+                    mayDumpDaemonLog(cfg.daemonLog());
+                    throw new IOException("Failed to start daemon in time " + cfg.autostartDuration()
+                            + "; check daemon logs in " + cfg.daemonLog());
+                }
+                logger.debug("... waiting");
+                Thread.sleep(500);
+            }
+        } catch (InterruptedException e) {
+            throw new IOException("Interrupted", e);
+        }
+        if (!Files.exists(cfg.socketPath())) {
+            mayDumpDaemonLog(cfg.daemonLog());
+            throw new IOException("Failed to start daemon; check daemon logs in " + cfg.daemonLog());
         }
     }
 }

--- a/node/file/src/main/java/eu/maveniverse/maven/mimir/node/file/FileNodeFactory.java
+++ b/node/file/src/main/java/eu/maveniverse/maven/mimir/node/file/FileNodeFactory.java
@@ -26,16 +26,13 @@ import org.eclipse.aether.spi.connector.checksum.ChecksumAlgorithmFactory;
 public final class FileNodeFactory implements SystemNodeFactory {
     private final Map<String, KeyResolverFactory> keyResolverFactories;
     private final Map<String, ChecksumAlgorithmFactory> checksumFactories;
-    private final DirectoryLocker directoryLocker;
 
     @Inject
     public FileNodeFactory(
             Map<String, KeyResolverFactory> keyResolverFactories,
-            Map<String, ChecksumAlgorithmFactory> checksumFactories,
-            DirectoryLocker directoryLocker) {
+            Map<String, ChecksumAlgorithmFactory> checksumFactories) {
         this.keyResolverFactories = requireNonNull(keyResolverFactories, "keyResolverFactories");
         this.checksumFactories = requireNonNull(checksumFactories, "checksumFactories");
-        this.directoryLocker = requireNonNull(directoryLocker, "directoryLocker");
     }
 
     @Override
@@ -64,6 +61,6 @@ public final class FileNodeFactory implements SystemNodeFactory {
                 keyResolver,
                 fileNodeConfig.checksumAlgorithms(),
                 checksumFactories,
-                directoryLocker);
+                DirectoryLocker.INSTANCE);
     }
 }

--- a/node/file/src/test/java/eu/maveniverse/maven/mimir/node/file/FileNodeTest.java
+++ b/node/file/src/test/java/eu/maveniverse/maven/mimir/node/file/FileNodeTest.java
@@ -11,7 +11,6 @@ import eu.maveniverse.maven.mimir.shared.impl.naming.SimpleKeyMapperFactory;
 import eu.maveniverse.maven.mimir.shared.impl.naming.SimpleKeyResolverFactory;
 import eu.maveniverse.maven.mimir.shared.naming.KeyMapper;
 import eu.maveniverse.maven.mimir.shared.node.SystemEntry;
-import eu.maveniverse.maven.shared.core.fs.DirectoryLocker;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -45,8 +44,7 @@ public class FileNodeTest {
                                 Sha1ChecksumAlgorithmFactory.NAME,
                                 new Sha1ChecksumAlgorithmFactory(),
                                 Sha512ChecksumAlgorithmFactory.NAME,
-                                new Sha512ChecksumAlgorithmFactory()),
-                        new DirectoryLocker())
+                                new Sha512ChecksumAlgorithmFactory()))
                 .createNode(config)) {
             Optional<FileEntry> entry = fileNode.locate(keyMapper.apply(central, junit));
             assertFalse(entry.isPresent());
@@ -88,8 +86,7 @@ public class FileNodeTest {
                                 Sha1ChecksumAlgorithmFactory.NAME,
                                 new Sha1ChecksumAlgorithmFactory(),
                                 Sha512ChecksumAlgorithmFactory.NAME,
-                                new Sha512ChecksumAlgorithmFactory()),
-                        new DirectoryLocker())
+                                new Sha512ChecksumAlgorithmFactory()))
                 .createNode(config)) {
             Optional<FileEntry> entry = fileNode.locate(keyMapper.apply(central, junit));
             assertFalse(entry.isPresent());
@@ -131,8 +128,7 @@ public class FileNodeTest {
                         Sha1ChecksumAlgorithmFactory.NAME,
                         new Sha1ChecksumAlgorithmFactory(),
                         Sha512ChecksumAlgorithmFactory.NAME,
-                        new Sha512ChecksumAlgorithmFactory()),
-                new DirectoryLocker());
+                        new Sha512ChecksumAlgorithmFactory()));
         try (FileNode fileNode1 = fileNodeFactory.createNode(config);
                 FileNode fileNode2 = fileNodeFactory.createNode(config)) {
             // should be ok
@@ -151,8 +147,7 @@ public class FileNodeTest {
                         Sha1ChecksumAlgorithmFactory.NAME,
                         new Sha1ChecksumAlgorithmFactory(),
                         Sha512ChecksumAlgorithmFactory.NAME,
-                        new Sha512ChecksumAlgorithmFactory()),
-                new DirectoryLocker());
+                        new Sha512ChecksumAlgorithmFactory()));
         try (FileNode fileNode = fileNodeFactory.createNode(config)) {
             assertThrows(IOException.class, () -> fileNodeFactory.createNode(config));
         }

--- a/node/jgroups/src/test/java/eu/maveniverse/maven/mimir/jgroups/JGroupsNodeTest.java
+++ b/node/jgroups/src/test/java/eu/maveniverse/maven/mimir/jgroups/JGroupsNodeTest.java
@@ -52,7 +52,7 @@ public class JGroupsNodeTest {
                 new SimpleKeyResolverFactory().createKeyResolver(config),
                 List.of(Sha1ChecksumAlgorithmFactory.NAME),
                 Map.of(Sha1ChecksumAlgorithmFactory.NAME, new Sha1ChecksumAlgorithmFactory()),
-                new DirectoryLocker());
+                DirectoryLocker.INSTANCE);
         FileNodeConfig configTwo = FileNodeConfig.of(
                 two, true, Collections.singletonList("SHA-1"), SimpleKeyResolverFactory.NAME, false, false);
         FileNode nodeTwo = new FileNode(
@@ -63,7 +63,7 @@ public class JGroupsNodeTest {
                 new SimpleKeyResolverFactory().createKeyResolver(config),
                 List.of(Sha1ChecksumAlgorithmFactory.NAME),
                 Map.of(Sha1ChecksumAlgorithmFactory.NAME, new Sha1ChecksumAlgorithmFactory()),
-                new DirectoryLocker());
+                DirectoryLocker.INSTANCE);
 
         String testGroup = UUID.randomUUID().toString();
         JChannel channelOne = new JChannel("udp-new.xml")

--- a/pom.xml
+++ b/pom.xml
@@ -71,7 +71,7 @@
     <requireRuntimeMavenVersion.range>[3.9,)</requireRuntimeMavenVersion.range>
 
     <!-- Dependency versions -->
-    <version.maveniverseShared>0.1.4</version.maveniverseShared>
+    <version.maveniverseShared>0.1.5</version.maveniverseShared>
     <version.maven>3.9.9</version.maven>
     <version.resolver>1.9.23</version.resolver>
     <version.slf4j>2.0.17</version.slf4j>

